tools/release.go

```diff
@@ -0,0 +1,154 @@
+package main
+
+// release.go creates a new version tag and pushes it to GitHub. It has two
+// main modes of operation:
+// * default: fetch origin/main, tag that commit with Major.Minor.Patch version
+//   of 0.YYYYMMDD.0, and push the tag.
+// * hotfix: fetch the named release tag, create a new branch starting from
+//   that commit, cherry-pick the named commit(s) onto that branch, tag the
+//   new branch head with a Major.Minor.Patch that matches the Major.Minor of
+//   named release tag and a Patch that is one greater, and push that tag.
+
+import (
+	"flag"
+	"fmt"
+	"os"
+	"os/exec"
+	"strconv"
+	"strings"
+	"time"
+
+	"golang.org/x/mod/semver"
+)
+
+func runOrDie(cmd *exec.Cmd) string {
+	fmt.Println(cmd.String())
+	out, err := cmd.CombinedOutput()
+	if err != nil {
+		fmt.Print(string(out))
+		fmt.Println(err)
+		os.Exit(1)
+	}
+	return string(out)
+}
+
+func main() {
+	if len(os.Args) >= 2 && os.Args[1] == "hotfix" {
+		os.Args = append([]string{os.Args[0]}, os.Args[2:]...)
+		hotfix()
+	} else {
+		release()
+	}
+}
+
+func release() {
+	var push bool
+	flag.BoolVar(&push, "f", false, "If set, push the resulting release tag to GitHub.")
+	flag.Parse()
+
+	// Fetch all of the latest commits and refs from origin, so that we can ensure
+	// we're tagging the correct commit and not recreating an existing tag.
+	runOrDie(exec.Command("git", "fetch", "origin"))
+
+	// We use semver's vMajor.Minor.Patch format, where the Major version is
+	// always 0 (no backwards compatibility guarantees), the Minor version is
+	// the date of the release, and the Patch number is zero for normal releases
+	// and only non-zero for hotfix releases.
+	minor := time.Now().Format("20060102")
+	version := fmt.Sprintf("v0.%s.0", minor)
+	message := fmt.Sprintf("Release %s", version)
+
+	// Produce the tag, using -s to PGP sign it. This will fail if a tag with
+	// that name already exists.
+	runOrDie(exec.Command("git", "tag", "-s", "-m", message, version, "origin/main"))
```

Comment:
```suggestion
	runOrDie(exec.Command("git", "tag", "-s", "-m", "'"+message+"'", version, "origin/main"))
```

Reply:
One of the things that makes the `exec` package cool is that this command isn't being run through a shell; it isn't concatenated into a single string and then executed. Instead, each of these arguments is passed whole to the underlying executable, so there's no need to quote any of these.

For a demonstration, compare https://go.dev/play/p/suwTOJnBp0I to the result of running `$ echo a            +              b` and the result of running `$ echo "a            +            b"` in a shell.

Reply:
This is still wild to me.